File: Monitor_matlab/STE.m
function [ feature_STE feature_STEAcc ] = STE( voice, fs, nBit )
%STE Calc Short-Time Energy of the voice
%   input: voice data
%   input: sample rate
%   input: number of bits per sample
%   output: short-time-energy feature of the voice data

    disp('Feature STE');
    frameLength = 20*0.001*fs; %frame length
    frameInc = frameLength;   %frame inc = frame length
    voiceFrames = enframe(voice, frameLength, frameInc);
    [framesNum frameLen] = size(voiceFrames); 
    
    temp = voiceFrames.^2;
    energy = sum(temp,2);
    
    feature_STE = mean(energy(energy>=(0.2*max(energy))));
    
    acc = diff(energy);
    feature_STEAcc = max(acc);
%     sortedAcc = sort(acc,'descend');
%     cutting = floor(length(sortedAcc)/2);
%     feature_STEAcc = mean(sortedAcc(1:cutting));
%     feature_STEAcc = mean(acc(acc>=(0.5*max(acc))));
    
%     plot(energy);
%     xlabel('Frame Num');
%     ylabel('Energy');
%     legend('Short-time Energy');

end

File: Monitor_c/STE.c
#include "STE.h"
#include "max.h"
#include "mean.h"
#include <malloc.h>

void STE(double data[], int size, int fs, int nBit, double *feature_STE, double *feature_STEAcc)
{
	int i,k;
	int max_pos;
	double maxEnergy, threshold;
	double secPerFra = 20*0.001;
	int frameLength = secPerFra*fs;
	int frameInc = frameLength;
	double **frames;
	double *energy; 
	double *energyDiff;
	int frameNum;	
	double energySum = 0;
	double energyBiggerNum = 0;

	enframe(data, size, frameLength, frameInc, &frames, &frameNum);
	energy =(double *)malloc(frameNum*sizeof(double));
	energyDiff =(double *)malloc(frameNum*sizeof(double));

	for(i=0; i<frameNum; i++)
	{
		energy[i] = 0;
		for(k=0; k<frameLength; k++)
			energy[i] += frames[i][k]*frames[i][k];
	}			

	maxEnergy = max(energy,frameNum,1,-1,&max_pos);
	threshold = 0.2*maxEnergy;

	for(i=0; i<frameNum; i++)
	{
		if (energy[i]>threshold)
		{
			energySum += energy[i];
			energyBiggerNum++;
		}
	}

	*feature_STE = energySum/energyBiggerNum;

	for(i=0; i<frameNum-1; i++)
		energyDiff[i] = energy[i+1]-energy[i];
	energyDiff[frameNum-1] = 0;

	*feature_STEAcc = max(energyDiff,frameNum,1,-1,&max_pos);

	free(energy);
	free(energyDiff);
}
